### DISCUSSION TITLE
MOOSE seems unable to recognize 1D boundary on 3D mesh

### DISCUSSION THREAD

**Discussion by richmondodufisan** [idaholab/moose#28869] [Q&A Meshing]
Hi all,

I have a problem which I have encountered in the past that is coming up again. I use gmsh for meshing.

When I create a 2D mesh, and define 1D physical groups (lines), MOOSE is able to recognize those lines as boundaries and do things like apply Neumann conditions, use VectorPostProcessors like SideValueSampler, etc.

However, when I create a 3D mesh and define 1D physical groups, I get the error:

```
(VectorPostprocessors/upper_right_z_disp/boundary):
    the following side sets (ids) do not exist on the mesh: sample_location (19)

    MOOSE distinguishes between "node sets" and "side sets" depending on whether
    you are using "Nodal" or "Integrated" BCs respectively. Node sets corresponding
    to your side sets are constructed for you by default.

    Try setting "Mesh/construct_side_list_from_node_list=true" if you see this error.
    Note: If you are running with adaptivity you should prefer using side sets.
```

This only shows up for 1D physical groups. 2D physical groups/boundaries work fine. Also, "construct_side_list_from_node_list=true" does not change anything.

Based off of this behavior, If I understand correctly, the way it works is that "sides" can only be 1 dimension less than the dimension of the mesh. But then what do I do if I want to apply a line load on a 3D mesh? Or in my case (reason for making this post), sample a value along a curved line? LineValueSampler can't work for me there because that can only define straight lines.

Interestingly enough, a 0D physical group had no issues applying a Dirichlet BC.

I saw this answer https://github.com/idaholab/moose/discussions/24174 from @jwpeterson , where it was suggested to add "lower_dimensional_block" to my mesh file, so I went ahead and did:

```
$MeshFormat
4.1 0 8
$EndMeshFormat
$PhysicalNames
2
0 18 "loading_point"
1 19 "sample_location" lower_dimensional_block
$EndPhysicalNames
$Entities
7 12 8 1
1 0 0 0 0 
2 0 0.002 0 0 
3 0 -0.002 0 0 
4 0 0 0.002 1 18 
5 0 0 -0.002 0 
6 0.002 0 0 0 
7 -0.002 0 0 0 
1 -6.389406329081476e-05 -6.389406329081475e-05 -6.389406329081475e-05 6.389406329081476e-05 0.002063894063290815 0.002063894063290815 0 2 4 -2 
2 -2.334153974494834e-05 -2.334153974494803e-05 -0.002023341539744949 2.334153974494834e-05 0.002023341539744949 2.334153974494846e-05 0 2 2 -5 

```

but nothing changed. How can I fix this so I can sample points along a curved line in 3D? I attached here my gmsh file, a simple sphere, and I want to sample the value of the displacement along a curved line on the surface from the top to the right side.

```
// Gmsh project created on Wed Oct 16 19:47:43 2024
SetFactory("OpenCASCADE");
//+
Point(1) = {0, 0, 0, 0.00005};
//+
Point(2) = {0, 0.002, 0, 0.00005};
//+
Point(3) = {0, -0.002, 0, 0.00005};
//+
Point(4) = {0, 0, 0.002, 0.00005};
//+
Point(5) = {0, 0, -0.002, 0.00005};
//+
Point(6) = {0.002, 0, 0, 0.00005};
//+
Point(7) = {-0.002, 0, 0, 0.00005};
//+
Circle(1) = {4, 1, 2};
//+
Circle(2) = {2, 1, 5};
//+
Circle(3) = {6, 1, 2};
//+
Circle(4) = {2, 1, 7};
//+
Circle(5) = {4, 1, 3};
//+
Circle(6) = {5, 1, 3};
//+
Circle(7) = {6, 1, 3};
//+
Circle(8) = {7, 1, 3};
//+
Circle(9) = {4, 1, 7};
//+
Circle(10) = {4, 1, 6};
//+
Circle(11) = {6, 1, 5};
//+
Circle(12) = {5, 1, 7};
//+
Curve Loop(2) = {1, -3, -10};
//+
Surface(1) = {2};
//+
Curve Loop(4) = {9, -4, -1};
//+
Surface(2) = {4};
//+
Curve Loop(6) = {2, -11, 3};
//+
Surface(3) = {6};
//+
Curve Loop(8) = {2, 12, -4};
//+
Surface(4) = {8};
//+
Curve Loop(10) = {5, -7, -10};
//+
Surface(5) = {10};
//+
Curve Loop(12) = {7, -6, -11};
//+
Surface(6) = {12};
//+
Curve Loop(14) = {6, -8, -12};
//+
Surface(7) = {14};
//+
Curve Loop(16) = {5, -8, -9};
//+
Surface(8) = {16};
//+
Surface Loop(1) = {7, 6, 5, 1, 2, 4, 3, 8};
//+
Volume(1) = {1};
//+
Surface Loop(2) = {5, 1, 2, 4, 3, 6, 7, 8};
//+
Physical Point("loading_point", 18) = {4};
//+
Physical Curve("sample_location", 19) = {10};


```

**Comment by GiudGiud** [comment]
Hello

I think from previous posts we ve seen that the gmsh reader does not support reading 1D sets inside a 3D mesh. You should try to make sure of this on a simple case

If your 1D nodeset is simple enough it should be possible at add it after loading the mesh in moose. 

Guillaume

**Reply by richmondodufisan** [reply]
What would I use to define it? It is fairly simple (y = 0; x^2 + z^2 = 0.002^2; z > 0) i.e a circle arc

**Reply by richmondodufisan** [reply]
Also would like to know if there is any open source mesh generator that can do this because it is a bit of a common problem (distributed line loads, needing to output values along a (custom) line, etc)

**Reply by GiudGiud** [reply]
We can definitely generate that.
I ll take a look. 
We might just need a ParsedGenerateNodeset

**Reply by GiudGiud** [reply]
I just made this ParsedGenerateNodesetobject which you can use from this branch 
https://github.com/idaholab/moose/pull/28873

The expression you have can be used to select nodes from the mesh that match it

I have not tested it yet so please report on any difficulty you encounter. It compiles that's all I checked for now

**Reply by richmondodufisan** [reply]
Can I place it in myapp/src/kernels (and myapp/include/kernels) ? I did that, and changed this line

`registerMooseObject("purpleApp", ParsedGenerateNodeset);`

but it didn't compile. I'm asking because I also use MOOSE on an HPC and Installing/updating is a real headache for the HPC managers due to the way it's setup (something about needing to use a 'singularity' container or something), and so we only update once a year (and I just updated a few weeks ago).

**Reply by GiudGiud** [reply]
You would need to bring the base class as well to do that. 

So 4 files in total, two headers and two source files

**Reply by richmondodufisan** [reply]
Ah gotcha. I did that now and it compiled. Here is the block in my input file:

```
  [output_sampling_line]
    type = ParsedGenerateNodeset
	input = sample_mesh
	combinatorial_geometry = '(y > 0-1e-8) & (y < 0+1e-8) & ((x^2 + z^2 - 0.002^2) > 0-1e-8) & ((x^2 + z^2 - 0.002^2) < 0+1e-8) & (z > 0-1e-8) & (z < 0+1e-8)'
	new_sideset_name = sample_line
 []
```

And here is the error I got:

```
*** ERROR ***
Unable to find a parameter with name: constant_expression when adding to group Parsed expression.

Stack frames: 23
0: libMesh::print_trace(std::ostream&)
1: moose::internal::mooseErrorRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
2: void mooseError<char const (&) [39], std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char>(char const (&) [39], std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char&&)
3: InputParameters::addParamNamesToGroup(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
4: ParsedGenerateNodeset::validParams()
5: RegistryEntry<ParsedGenerateNodeset>::buildParameters()
6: Factory::getValidParams(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const
7: MooseObjectAction::MooseObjectAction(InputParameters const&)
8: AddMeshGeneratorAction::AddMeshGeneratorAction(InputParameters const&)
9: RegistryEntry<AddMeshGeneratorAction>::buildAction(InputParameters const&)
10: ActionFactory::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, InputParameters&)
11: Moose::Builder::walkRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, hit::Node*)
12: Moose::Builder::walk(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, hit::Node*)
13: hit::Node::walk(hit::Walker*, hit::NodeType, hit::TraversalOrder)
14: hit::Node::walk(hit::Walker*, hit::NodeType, hit::TraversalOrder)
15: hit::Node::walk(hit::Walker*, hit::NodeType, hit::TraversalOrder)
16: Moose::Builder::build()
17: MooseApp::setupOptions()
18: MooseApp::run()
19: main
20: /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7fa84ae44d90]
21: __libc_start_main
22: ../purple-opt(+0x33c7) [0x55b1458e93c7]
Abort(1) on node 0 (rank 0 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
```

**Reply by GiudGiud** [reply]
Ok I ll fix that soon. The parameter group (bottom of validParams) must have a typo if you want to fix it

**Reply by richmondodufisan** [reply]
Fixed that, there was another error too (changed `"expression"` to `"combinatorial_geometry"` in all places requested) and it worked! Thanks. 

But there's another issue- the mesh happened to only have nodes at the start and end of those locations, so the csv output only had two datapoints. Would it be possible to define how many new nodes you want to place along that line (say if you wanted to apply a distributed line load)? Or in my case where I just need the solution, to "sample" at a number of locations defined by the user?

**Reply by GiudGiud** [reply]
this is probably the "use external only" parameter doing that
try stting it to false

**Reply by GiudGiud** [reply]
if it's not that, then maybe add a tolerance to the expression, make it a little looser. Instead of a>0, do a>-epsilon

**Reply by richmondodufisan** [reply]
I don't think there is a "use_external_only" parameter

```
*** ERROR ***
/home/richmond98/projects/purple/Cornea_Linear_Elastic/Cornea_Harmonic_Perturbation_3D.i:26.2: unused parameter 'Mesh/output_sampling_line/use_external_only'


Stack frames: 11
0: libMesh::print_trace(std::ostream&)
1: moose::internal::mooseErrorRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
2: void mooseError<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)
3: Moose::Builder::errorCheck(libMesh::Parallel::Communicator const&, bool, bool)
4: MooseApp::errorCheck()
5: MooseApp::executeExecutioner()
6: MooseApp::run()
7: main
8: /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7fdb9ff52d90]
9: __libc_start_main
10: ../purple-opt(+0x33c7) [0x56460882f3c7]
Abort(1) on node 0 (rank 0 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
```

and I used a tolerance of 1e-8, trying to keep it as close to the line as possible. Increased to 1e-6 with no change too.

But what I would really like to do is to add more nodes, almost like a way of refining the mesh at those points. And if not possible, add more 'points' to interpolate the solution for the output like it's done in LineValueSampler.

**Reply by richmondodufisan** [reply]
Actually, looking at the mesh closer, there are more nodes along that path. More than just the two it found.

**Reply by GiudGiud** [reply]
> I don't think there is a "use_external_only" parameter

not literally but there is an option named similarly

> and I used a tolerance of 1e-8, trying to keep it as close to the line as possible. Increased to 1e-6 with no change too.

I m not sure how much of a tolerance you need in case of a polygonization effect for meshing a circle for example

**Reply by richmondodufisan** [reply]
gotcha. it was called "include_only_external_nodes". I set it to false in my input file but I think it was already false by default. No change in the output CSV. Also my mesh dimensions are on the order of radius = 2e-3 (a sphere with triangular elements) so any lower tolerance will likely include unwanted nodes.

**Reply by richmondodufisan** [reply]
@GiudGiud Hi, I wanted to know if you have any update on this/have any alternative I can use in the meantime. Specifically to be able to plot over the curved line. I'm trying to see if paraview can do this but it seems i'll have to do some scripting.

**Reply by GiudGiud** [reply]
For plotting over the curved line you sill need to create / use a Positions object that represents the curved line. 
If you have a nodeset already for that line it should be easy

Then use the VPP that works with positions to sample any functor (variables, functions and functor material properties) on those positions

**Reply by richmondodufisan** [reply]
Yeah that's the problem, 1D nodesets I implemented in gmsh are not recognizable by MOOSE for some reason. Hence why I was hoping the ParsedGenerateNodeset you helped create would work and simplify this process. 

The only option I'm looking at now seems to be to make some kind of script that tries to generate a whole bunch of nodal positions to output the data either in MOOSE or with paraview.

**Reply by GiudGiud** [reply]
Ok so you are still generating this 1D nodeset. 

Given your simple expression for the nodeset I think ParsedNodesetGrnerator would work. 

If you cannot get it to work please share your mesh so I can investigate why

**Reply by richmondodufisan** [reply]
Sure, here's the mesh (had to zip to upload here). Thank you!

[eyeball_3D_coarse.zip](https://github.com/user-attachments/files/17479826/eyeball_3D_coarse.zip)

Also here in case you can't use zip files (finer mesh): https://github.com/richmondodufisan/purple/blob/main/Cornea_Linear_Elastic/eyeball_3D.msh

**Reply by GiudGiud** [reply]
Thanks! I ll have time later this week

**Reply by GiudGiud** [reply]
does this work?
```
[Mesh]
    [fmg]
        type = FileMeshGenerator
        file = eyeball_3D_coarse.msh 
    []
    [nodes]
        type = ParsedGenerateNodeset
        input = 'fmg'
        expression = '(abs(y) < 1e-8) & (abs(x * x + z * z - 0.002 *0.002) < 1e-8 ) & (z > 0)'
        new_nodeset_name = 'circle'
    []
[]
```
I get a half circle of nodes on the outside

**Reply by richmondodufisan** [reply]
Yes that does! Thank you so much!! That is weird though, why would it work with "abs" and not with doing them individually?

**Reply by GiudGiud** [reply]
abs lets us specify tolerances easier
